porch/repository/pkg/git/testing.go

```diff
@@ -0,0 +1,635 @@
+		} else {
+			klog.Warningf("updated reference %v -> %v", refUpdate.Ref, refUpdate.To)
+		}
+	}
+
+	return nil
+}
+
+// objectWalker is based on objectWalker in go-git/v5
+
+type objectWalker struct {
+	Storer storage.Storer
+	// seen is the set of objects seen in the repo.
+	// seen map can become huge if walking over large
+	// repos. Thus using struct{} as the value type.
+	seen map[plumbing.Hash]struct{}
+}
+
+func newObjectWalker(s storage.Storer) *objectWalker {
+	return &objectWalker{s, map[plumbing.Hash]struct{}{}}
+}
+
+// walkAllRefs walks all (hash) references from the repo.
+func (p *objectWalker) walkAllRefs() error {
+	// Walk over all the references in the repo.
+	it, err := p.Storer.IterReferences()
+	if err != nil {
+		return err
+	}
+	defer it.Close()
+	err = it.ForEach(func(ref *plumbing.Reference) error {
+		// Exit this iteration early for non-hash references.
+		if ref.Type() != plumbing.HashReference {
+			return nil
+		}
+		return p.walkObjectTree(ref.Hash())
+	})
+	return err
+}
+
+func (p *objectWalker) isSeen(hash plumbing.Hash) bool {
+	_, seen := p.seen[hash]
+	return seen
+}
+
+func (p *objectWalker) add(hash plumbing.Hash) {
+	p.seen[hash] = struct{}{}
+}
+
+// walkObjectTree walks over all objects and remembers references
+// to them in the objectWalker. This is used instead of the revlist
+// walks because memory usage is tight with huge repos.
+func (p *objectWalker) walkObjectTree(hash plumbing.Hash) error {
+	// Check if we have already seen, and mark this object
+	if p.isSeen(hash) {
+		return nil
+	}
+	p.add(hash)
+	// Fetch the object.
+	obj, err := object.GetObject(p.Storer, hash)
+	if err != nil {
+		return fmt.Errorf("Getting object %s failed: %v", hash, err)
+	}
+	// Walk all children depending on object type.
+	switch obj := obj.(type) {
+	case *object.Commit:
+		err = p.walkObjectTree(obj.TreeHash)
+		if err != nil {
+			return err
+		}
+		for _, h := range obj.ParentHashes {
+			err = p.walkObjectTree(h)
+			if err != nil {
+				return err
+			}
+		}
+	case *object.Tree:
+	nextEntry:
+		for i := range obj.Entries {
+			switch obj.Entries[i].Mode {
+			case filemode.Executable, filemode.Regular, filemode.Symlink:
+				p.add(obj.Entries[i].Hash)
+				continue nextEntry
+			case filemode.Submodule:
+				// hash is the submodule ref, I believe
+				continue nextEntry
+			case filemode.Dir:
+				// process recursively
+			default:
+				klog.Warningf("unknown entry mode %s", obj.Entries[i].Mode)
+			}
+			// Normal walk for sub-trees (and symlinks etc).
+			err = p.walkObjectTree(obj.Entries[i].Hash)
+			if err != nil {
+				return err
+			}
+		}
+	case *object.Tag:
+		return p.walkObjectTree(obj.Target)
+	default:
+		// Error out on unhandled object types.
+		return fmt.Errorf("Unknown object %s %s %T\n", obj.ID(), obj.Type(), obj)
+	}
+	return nil
+}
+
+// initRepo is a helper that creates a first commit, ensuring the repo is not empty.
+func initRepo(repo *git.Repository) error {
```

Comment:
TBD, but we might want this in a different package in the end - depends on whether we further split "git server" from "pure test code"

Reply:
great point. I noticed this after I'd made the initial split. I'll refactor this in subsequent PRs.

We'll also want different ways to initialize the repositories so I'll have couple of initializers (empty repo, initialized from contents in testdata, ...)